integrators/simple_harmonic_motion.cpp

```diff
@@ -0,0 +1,363 @@
+﻿
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "geometry/sign.hpp"
+#include "glog/logging.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "integrators/symplectic_partitioned_runge_kutta_integrator.hpp"
+#include "quantities/quantities.hpp"
+#include "quantities/named_quantities.hpp"
+#include "testing_utilities/almost_equals.hpp"
+#include "testing_utilities/numerical_analysis.hpp"
+#include "testing_utilities/numerics.hpp"
+#include "testing_utilities/statistics.hpp"
+
+namespace principia {
+
+using quantities::Abs;
+using quantities::Acceleration;
+using quantities::AngularFrequency;
+using quantities::Energy;
+using quantities::Force;
+using quantities::Length;
+using quantities::Mass;
+using quantities::Speed;
+using quantities::Pow;
+using quantities::Power;
+using quantities::SIUnit;
+using quantities::Speed;
+using quantities::Stiffness;
+using quantities::Time;
+using si::Joule;
+using si::Kilogram;
+using si::Metre;
+using si::Newton;
+using si::Second;
+using testing_utilities::AlmostEquals;
+using testing_utilities::BidimensionalDatasetMathematicaInput;
+using testing_utilities::ComputeHarmonicOscillatorAcceleration;
+using testing_utilities::ComputeHarmonicOscillatorForce;
+using testing_utilities::ComputeHarmonicOscillatorVelocity;
+using testing_utilities::PearsonProductMomentCorrelationCoefficient;
+using testing_utilities::RelativeError;
+using testing_utilities::Slope;
+using ::testing::AllOf;
+using ::testing::Eq;
+using ::testing::Gt;
+using ::testing::Lt;
+using ::testing::Ne;
+using ::testing::ValuesIn;
+
+#define INTEGRATOR(name) &name(), #name
+
+namespace integrators {
+
+namespace {
+
+struct SimpleHarmonicMotionTestableProperties {
+  not_null<SRKNIntegrator const*> integrator;
+  std::string name;
+  int convergence_order;
+  // Convergence, in the sense tested below, of the above order occurs for
+  // timesteps smaller than |beginning_of_convergence| when integrating the unit
+  // harmonic oscillator.
+  Time beginning_of_convergence;
+  // The expected errors when integrating the unit harmonic oscillator for
+  // 1000 s with a 1 ms timestep.
+  Length expected_position_error;
+  Speed expected_velocity_error;
+  // The expected errors when integrating the unit harmonic oscillator for
+  // 500 s with a 0.2 s timestep.
+  Energy expected_energy_error;
+};
+
+// This allows the test output to be legible, i.e.,
+// "where GetParam() = Leapfrog" rather than
+// "where GetParam() = n-byte object <hex>"
+std::ostream& operator<<(std::ostream& stream,
+                         SimpleHarmonicMotionTestableProperties param) {
+  return stream << param.name;
+}
+
+std::vector<SimpleHarmonicMotionTestableProperties> Instances() {
+  return {
+      {INTEGRATOR(Leapfrog), 2, 0.4 * Second,
```

Comment:
I'd prefer a line break systematically after `INTEGRATOR(...),` to make sure that all entries look the same.